### PR TITLE
Feature/280 secret 3

### LINKED ISSUE #280: Add support for reading secrets from AWS Secrets Manager
# Background
Secrets should not be stored as plaintext in the configuration, nor in files on hdfs or s3. They should be retrieved directly by the Spark job from a vault. Furthermore, secrets' values must not be printed in the logs. As a first vault provider, AWS Secrets Manager should be supported.

# Suggestions

Config (mental model, if we had yaml config)
```
secretsprovider:
  secrets:
    truststorepassword:
     options:
       secretname: <the-secret-name>
       readasmap: true
       key: <the-key>
       encoding: base64
     secretvalue: // should be filled by provider
    otherpassword:
     provider: dummymanager
     options:
       dummykey: <dummy-value>
     secretvalue: // should be filled by provider

  config:
    providers:
      awssecretsmanager:
        class: za.co.absa.hyperdrive.secretsprovider.AwsSecretsManager
      dummymanager:
        class: za.co.absa.hyperdrive.dummymanager
    defaultprovider: awssecretsmanager
reader:
 option:
  kafka:
     ssl.truststore.password: ${secrets.truststorepassword.value}
```

A secret that should be retrieved by aws secrets manager would be configured as follows:

```
secretsprovider.config.providers.awssecretsmanager.class=za.co.absa.hyperdrive.secretsprovider.AwsSecretsManager
secretsprovider.secrets.truststorepassword.provider=awssecretsmanager
secretsprovider.secrets.truststorepassword.options.secretname=<the-secret-name>
secretsprovider.secrets.truststorepassword.options.key=<the-secret-key>
secretsprovider.secrets.truststorepassword.options.encoding=base64
reader.option.kafka.ssl.truststore.password=${secretsprovider.secrets.truststorepassword.secretvalue}
```

In this example, both `awssecretsmanager` and `truststorepassword` are arbitrary identifiers that can be chosen when the configuration is written.

Before the configuration is resolved, all secrets must be resolved by the provider, given the `options` for each secret. Global config for the provider is given by the key-value map under `secrets.config.providers.[provider-id].options`

### PR DESCRIPTION
**Pull request by kevinwallimann**
Closes #280

### CODE REVIEW

File: api/src/test/scala/za/co/absa/hyperdrive/ingestor/api/utils/TestConfigUtils.scala [hunk 1]
```diff
@@ -386,4 +386,43 @@ class TestConfigUtils extends FlatSpec with Matchers with MockitoSugar {
...
+    ConfigurationConverter.getMap(subsets("myItem2")).asScala should contain theSameElementsAs Map(
+      "name" -> "My Item 2",
+      "description" -> "Very nice item 2",
+      "options.color" -> "blue",
+      "options.size" -> "large"
+    )
+  }
+
+  "getSubsets" should "return the correct subsets if the parent key ends with a dot" in {
+    val config = new BaseConfiguration
```
Comment (jozefbakus) [inline api/src/test/scala/za/co/absa/hyperdrive/ingestor/api/utils/TestConfigUtils.scala:418]: I would add items that does not match similar to the previous test.

File: api/src/test/scala/za/co/absa/hyperdrive/ingestor/api/utils/TestConfigUtils.scala [hunk 2]
```diff
@@ -386,4 +386,43 @@ class TestConfigUtils extends FlatSpec with Matchers with MockitoSugar {
...
+    config.addProperty("service.items.myItem1.name", "My Item 1")
+
+    val subsets = ConfigUtils.getSubsets(config, "service.items.")
+
+    import scala.collection.JavaConverters._
+    subsets.keys should contain theSameElementsAs Array("myItem1")
+    subsets("myItem1").getKeys.asScala.toSeq should contain theSameElementsAs Array("name")
+    subsets("myItem1").getString("name") shouldBe "My Item 1"
+  }
```
Comment (jozefbakus) [inline api/src/test/scala/za/co/absa/hyperdrive/ingestor/api/utils/TestConfigUtils.scala:436]: Add one more test where there is no match and it returns empty map

File: parent-conf/pom.xml [hunk 3]
```diff
@@ -218,6 +219,13 @@
                 <artifactId>commons-beanutils</artifactId>
                 <version>${commons.beanutils.version}</version>
             </dependency>
+            <dependency>
+                <groupId>software.amazon.awssdk</groupId>
+                <artifactId>bom</artifactId>
+                <version>${aws.java.sdk.bom.version}</version>
+                <type>pom</type>
```
Comment (jozefbakus) [inline parent-conf/pom.xml:226]: Wouldn't be better to just have  aws-java-sdk-secretsmanager instead of having really long list of deps from this one ?

File: api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/secrets/SecretsProvider.scala [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+package za.co.absa.hyperdrive.ingestor.api.secrets
+
+import org.apache.commons.configuration2.Configuration
+
+trait SecretsProvider {
+  def retrieveSecret(options: Configuration): String
+}
+
+object SecretsProvider {
+  val BaseKey = "secretsprovider"
```
Comment (jozefbakus) [inline api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/secrets/SecretsProvider.scala]: This is "scala" way to have uppercase first letter. But in our classes we use lowercase first letter.

File: api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/secrets/SecretsProvider.scala [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+package za.co.absa.hyperdrive.ingestor.api.secrets
+
+import org.apache.commons.configuration2.Configuration
+
+trait SecretsProvider {
+  def retrieveSecret(options: Configuration): String
+}
+
+object SecretsProvider {
```
Comment (jozefbakus) [inline api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/secrets/SecretsProvider.scala]: Writer, Readers and transformers have attributes in XxxxAttributes class

File: driver/src/main/scala/za/co/absa/hyperdrive/driver/IngestionDriver.scala [hunk 6]
```diff
@@ -29,11 +30,16 @@ private[driver] abstract class IngestionDriver {
...
   val ListDelimiter = ','
 
+
   def main(args: Array[String]): Unit = {
     logger.info(s"Starting Hyperdrive ${DriverUtil.getVersionString}")
     val configuration = loadConfiguration(args)
+    SecretsConfigUtils.resolveSecrets(configuration)
     logger.info("Configuration loaded.")
-    printConfiguration(configuration)
+    SecretsConfigUtils.getRedactedConfigurationAsMap(configuration)
```
Comment (jozefbakus) [inline driver/src/main/scala/za/co/absa/hyperdrive/driver/IngestionDriver.scala]: I would leave printing in separate method

File: driver/src/main/scala/za/co/absa/hyperdrive/driver/IngestionDriver.scala [hunk 7]
```diff
@@ -29,11 +30,16 @@ private[driver] abstract class IngestionDriver {
   private val logger = LoggerFactory.getLogger(this.getClass)
   val ListDelimiter = ','
 
+
   def main(args: Array[String]): Unit = {
     logger.info(s"Starting Hyperdrive ${DriverUtil.getVersionString}")
     val configuration = loadConfiguration(args)
+    SecretsConfigUtils.resolveSecrets(configuration)
```
Comment (jozefbakus) [inline driver/src/main/scala/za/co/absa/hyperdrive/driver/IngestionDriver.scala:37]: Oh, this is terrible, I would prefer to have it as a function but I think you can't do much about it

File: parent-conf/pom.xml (hunk 3 above)
Comment (kevinwallimann) [reply parent-conf/pom.xml:226]: It's just a bill of materials, only what you define under `dependencies` (not `dependencyManagement`) makes it into the final jar. Using the BOM is the recommended approach by the AWS documentation. Probably we could also manage the versions ourselves, by using a maven property like we do for the other dependencies. But why follow another process

File: api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/secrets/SecretsProvider.scala (hunk 5 above)
Comment (kevinwallimann) [reply api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/secrets/SecretsProvider.scala]: I moved it to a new object `SecretProviderCommonAttributes`

File: driver/src/main/scala/za/co/absa/hyperdrive/driver/IngestionDriver.scala (hunk 7 above)
Comment (kevinwallimann) [reply driver/src/main/scala/za/co/absa/hyperdrive/driver/IngestionDriver.scala:37]: I could `clone` the original config and then call the modifying methods on the clone and return the clone. Not sure if it's worth the effort. I could also just return a reference to the config, but then I pretend it's immutable when it's not.

File: driver/src/main/scala/za/co/absa/hyperdrive/driver/IngestionDriver.scala (hunk 6 above)
Comment (kevinwallimann) [reply driver/src/main/scala/za/co/absa/hyperdrive/driver/IngestionDriver.scala]: Done

File: api/src/test/scala/za/co/absa/hyperdrive/ingestor/api/utils/TestConfigUtils.scala (hunk 2 above)
Comment (kevinwallimann) [reply api/src/test/scala/za/co/absa/hyperdrive/ingestor/api/utils/TestConfigUtils.scala:436]: Done

File: api/src/test/scala/za/co/absa/hyperdrive/ingestor/api/utils/TestConfigUtils.scala (hunk 1 above)
Comment (kevinwallimann) [reply api/src/test/scala/za/co/absa/hyperdrive/ingestor/api/utils/TestConfigUtils.scala:418]: Done

File: api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/secrets/SecretsProvider.scala (hunk 4 above)
Comment (kevinwallimann) [reply api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/secrets/SecretsProvider.scala]: Done

File: driver/src/main/scala/za/co/absa/hyperdrive/driver/IngestionDriver.scala (hunk 7 above)
Comment (jozefbakus) [reply driver/src/main/scala/za/co/absa/hyperdrive/driver/IngestionDriver.scala:37]: I would leave it now it is more visible that it is mutable

**Review by jozefbakus — APPROVED** [review]
